--- src/ls_odbc.c.orig	2026-03-13 16:42:21.028485036 -0300
+++ src/ls_odbc.c	2026-03-13 16:42:21.036058999 -0300
@@ -219,7 +219,7 @@
 static param_data *malloc_stmt_params(SQLSMALLINT c)
 {
 	param_data *p = (param_data *)malloc(sizeof(param_data)*c);
-	memset(p, 0, sizeof(param_data)*c);
+	if (p) memset(p, 0, sizeof(param_data)*c);
 
 	return p;
 }
@@ -230,7 +230,10 @@
 		param_data *p = data;
 
 		for(; c>0; ++p, --c) {
-			free(p->buf);
+			if (p->buf) {
+				free(p->buf);
+				p->buf = NULL;
+			}
 		}
 		free(data);
 	}
@@ -372,7 +375,7 @@
 #if LUA_VERSION_NUM>=503
 		/* iNteger */
 		case 'n': {
-			SQLLEN num;
+			SQLINTEGER num;
 			SQLLEN got;
 			SQLRETURN rc = SQLGetData(hstmt, i, SQL_C_SLONG, &num, 0, &got);
 			if (error(rc))
@@ -380,7 +383,7 @@
 			if (got == SQL_NULL_DATA)
 				lua_pushnil(L);
 			else
-				lua_pushinteger(L, num);
+				lua_pushinteger(L, (lua_Integer)num);
 			return 0;
 		}
 #endif
@@ -729,10 +732,10 @@
 
 	if (numcols > 0) {
 		/* if there is a results table (e.g., SELECT) */
-		return create_cursor(L, -1, stmt, numcols);
+		return create_cursor(L, istmt, stmt, numcols);
 	} else {
 		/* if action has no results (e.g., UPDATE) */
-		SQLLEN numrows;
+		SQLLEN numrows = -1;
 		if(error(SQLRowCount(stmt->hstmt, &numrows))) {
 			return fail(L, hSTMT, stmt->hstmt);
 		}
