--- tests/test.lua	2026-03-12 12:02:38.026572410 -0300
+++ tests/test.lua.fixed	2026-03-12 12:02:38.023234703 -0300
@@ -792,8 +792,8 @@
 end
 
 datasource = arg[2] or DEFAULT_TEST_DATABASE or "luasql-test"
-username = arg[3] or DEFAULT_USERNAME or nil
-password = arg[4] or DEFAULT_PASSWORD or nil
+username = arg[3] or os.getenv("DB_USER") or DEFAULT_USERNAME or nil
+password = arg[4] or os.getenv("DB_PASS") or DEFAULT_PASSWORD or nil
 
 -- Complete set of tests
 tests = {
